libs/base58_nostd/src/lib.rs

```diff
@@ -0,0 +1,339 @@
+        );
+    }
+
+    #[test]
+    fn test_from_base58_invalid_char() {
+        assert!("0".from_base58().is_err());
+        assert!("O".from_base58().is_err());
+        assert!("I".from_base58().is_err());
+        assert!("l".from_base58().is_err());
+        assert!("3mJr0".from_base58().is_err());
+        assert!("O3yxU".from_base58().is_err());
+        assert!("3sNI".from_base58().is_err());
+        assert!("4kl8".from_base58().is_err());
+        assert!("s!5<".from_base58().is_err());
+        assert!("t$@mX<*".from_base58().is_err());
+    }
+
+    #[test]
+    fn test_from_base58_initial_zeros() {
+        assert_eq!("1ZiCa".from_base58().unwrap(), b"\0abc");
+        assert_eq!("11ZiCa".from_base58().unwrap(), b"\0\0abc");
+        assert_eq!("111ZiCa".from_base58().unwrap(), b"\0\0\0abc");
+        assert_eq!("1111ZiCa".from_base58().unwrap(), b"\0\0\0\0abc");
+    }
+
+    #[test]
+    fn test_to_base58_basic() {
+        assert_eq!(b"".to_base58(), "".as_bytes());
+        assert_eq!(&[32].to_base58(), "Z".as_bytes());
+        assert_eq!(&[45].to_base58(), "n".as_bytes());
+        assert_eq!(&[48].to_base58(), "q".as_bytes());
+        assert_eq!(&[49].to_base58(), "r".as_bytes());
+        assert_eq!(&[57].to_base58(), "z".as_bytes());
+        assert_eq!(&[45, 49].to_base58(), "4SU".as_bytes());
+        assert_eq!(&[49, 49].to_base58(), "4k8".as_bytes());
+        assert_eq!(b"abc".to_base58(), "ZiCa".as_bytes());
+        assert_eq!(b"1234598760".to_base58(), "3mJr7AoUXx2Wqd".as_bytes());
+        assert_eq!(
+            b"abcdefghijklmnopqrstuvwxyz".to_base58(),
+            "3yxU3u1igY8WkgtjK92fbJQCd4BZiiT1v25f".as_bytes()
+        );
+    }
+
+    #[test]
+    fn test_to_base58_initial_zeros() {
+        assert_eq!(b"\0abc".to_base58(), "1ZiCa".as_bytes());
+        assert_eq!(b"\0\0abc".to_base58(), "11ZiCa".as_bytes());
+        assert_eq!(b"\0\0\0abc".to_base58(), "111ZiCa".as_bytes());
+        assert_eq!(b"\0\0\0\0abc".to_base58(), "1111ZiCa".as_bytes());
+    }
+
+    #[test]
+    fn to_base58check() {
+        assert_eq!(
+            b"SOME_TOKEN_ID".to_mls_b58check(),
+            "25TfmUELb1jGfVSAbKsV4fAVTKAn".as_bytes()
+        );
+    }
+
+    #[test]
+    fn from_base58check() {
+        assert_eq!(
+            "25TfmUELb1jGfVSAbKsV4fAVTKAn".from_mls_b58check().unwrap(),
+            b"SOME_TOKEN_ID".to_vec()
+        );
+    }
+
+    #[test]
+    fn from_base58check_with_invalid_checksum() {
+        assert_eq!(
+            "j8YiVRUK8wrJ2wzLH7W6222".from_mls_b58check(),
+            Err(FromBase58Error::InvalidChecksum)
+        );
+    }
+
+    #[test]
+    #[should_panic]
+    fn from_base58check_with_invalid_length() {
+        "Wh4bh".from_mls_b58check().unwrap();
+    }
+
+    #[test]
+    fn base58check_loop() {
+        let text =
+            "To be, or not to be, that is the Question: Whether ’tis Nobler in the mind to suffer.";
+        let enc = text.as_bytes().to_mls_b58check();
```

Comment:
These tests are not enough. Here's my concern: If your encoding is incorrect for any reason, things can break widely.

Here's how the tests should be made:

1. Find a good library that's well tested (whether in python or C++ or whatever, doesn't matter)
2. Produce many results there from conversion of base58check
3. Use them here as chosen-texts as the bases for your tests, to make sure that your conversion is doing what it has to do

Step 1 can be made much, much better by simply using a library that's available in rust, but given we're very close to launch, we won't do this. You should find a library in rust that implements base58, and use that to produce arbitrary and random data as input with its results, and then compare the results with yours.

You mentioned that you're concerned that they use rust's std, which we cannot use: That shouldn't be a problem. Simply create another crate that can depend on std.

Reply:
Added new data from external projects:
- Data prepared from our functional tests.
- Js library https://github.com/wzbg/base58check/blob/master/test.js
- https://github.com/bitcoin/bitcoin/blob/master/src/test/base58_tests.cpp 

To be honest, in Bitcoin not many tests for b58check. Anyway, I didn't find any problems with our implementation. 

Done.

Reply:
We just need a few values that we can compare against. Bytes vs what bitcoin what give for them. Then we run the same function and ensure that our implementation and bitcoin's give the exact same result. You don't need a million cases.